lib/pbench/cli/server/tree_manage.py

```diff
@@ -0,0 +1,119 @@
+                    for link in path.iterdir():
+                        if tarball.name in link.name:
+                            states.append(name)
+                if states:
+                    states.sort()
+                    print(f"        States: {', '.join(states)}")
+
+
+@click.command()
+@pass_cli_context
+@click.option("--controller", help="Controller name for a created dataset")
+@click.option(
+    "--create",
+    type=click.Path(
+        exists=True, file_okay=True, dir_okay=False, readable=True, resolve_path=True
+    ),
+    prompt="Tarball path",
+    prompt_required=False,
+    help="Create a dataset from a tarball and MD5 file",
+)
+@click.option(
+    "--display", default=False, is_flag=True, help="Display the full tree on completion"
+)
+@click.option(
+    "--full/--no-full", default=False, help="Discover the full tree on startup"
+)
+@click.option("--user", help="Username to own a created dataset")
+@common_options
+def tree_manage(
+    context: object, controller: str, create: str, display: bool, full: bool, user: str
+):
+    """
+    Discover, display, and manipulate the on-disk representation of controllers
+    and datasets.
+
+    This primarily exposes the FileTree object hierarchy, and provides a simple
+    hierarchical display of controllers and datasets.
+
+    This command can also be used to create a dataset, using the Dataset and
+    FileTree classes to maintain equivalence with the PUT operation (skipping
+    the network upload when the tarball and MD5 file are available locally). We
+    don't implement `--delete` however, as that would require also integrating
+    with the Elasticsearch bulk delete: this has to be done through the API.
+
+    Args:
+        context: Click context (contains shared `--config` value)
+        controller: Controller name (required for `--create`)
+        create: Create a new on-disk dataset from a tarball path
+        display: Print a simplified representation of the hierarchy
+        full: Discover the full Pbench on-disk hierarchy
+        user: Username to own a new dataset
+    """
+    try:
+        config = config_setup(context)
+        logger = get_pbench_logger("filetree", config)
+
+        file_tree = FileTree(config, logger)
+        if full:
+            file_tree.full_discovery()
+
+        if create:
+            if not controller:
+                click.echo("Create requires a controller name", err=True)
+                exit(1)
+            if not user:
+                click.echo("Create requires a username", err=True)
+                exit(1)
+            owner = User.query(username=user)
+            dataset = Dataset.create(controller=controller, owner=owner, path=create)
+            tarball = file_tree.create(controller, Path(create))
+            Upload.finalize_dataset(dataset, tarball, config, logger)
+
+        if display:
+            print_tree(file_tree)
+
+        rv = 0
```

Comment:
It's true that I don't like the pointless-init-which-is-overwritten-later, and, similarly, I'm not a huge fan of the default-init-which-is-conditionally-overwritten-later, but I am a fan of "declaring variables" (and scoping), even though those aren't so much a Thing in Python.  So, I'm still trying to figure out what my preference should be in this crazy language.

What I want most to avoid (even more than silly initializations) is paths through the code, now or in the future, which result in a read of an undefined variable (i.e., a `NameError` which isn't detected until runtime).  So, having the variable "assigned on all paths" only works if the developer and the reviewers can trace _all_ the paths, and it works only until we create a new path....

But, I recognize the validity of your approach, and I appreciate the attempt to avoid pointless initializations...I just question the robustness of the practice.